src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs

```diff
@@ -0,0 +1,394 @@
+                }
+                return s_LocalFqdn;
+            }
+        }
+
+        // Installs the root certificate from the bridge 
+        // Needed for all tests so we trust the incoming certificate coming from the service/bridge
+        // We do our best to detect if we're running on the same box as the bridge; if so, we don't try to install the cert
+        // as that operation requires admin privileges
+        public static void InstallRootCertificateFromBridge()
+        {
+            // PUT the Authority to the Bridge (returns thumbprint)
+            var response = BridgeClient.MakeResourcePutRequest(CertificateAuthorityResourceName, null);
+
+            string thumbprint;
+            X509Certificate2 certificateToInstall = null;
+            bool rootCertificateAlreadyInstalled = false;
+
+            lock (s_certificateLock)
+            {
+                if (response.TryGetValue(ThumbprintKeyResourceString, out thumbprint))
+                {
+                    rootCertificateAlreadyInstalled = s_rootCertificates.ContainsKey(thumbprint);
+                }
+
+                if (rootCertificateAlreadyInstalled)
+                {
+                    // Cert's been installed already, bail out
+                    return;
+                }
+                else
+                {
+                    using (X509Store store = new X509Store(StoreName.Root, StoreLocation.LocalMachine))
+                    {
+                        store.Open(OpenFlags.ReadOnly);
+                        var collection = store.Certificates.Find(X509FindType.FindByThumbprint, thumbprint, false);
+                        if (collection.Count > 0)
+                        {
+                            // We don't need to actually add the cert ourselves, because this cert has previously been installed 
+                            // Likely BridgeClient is running on the same machine as the Bridge and the Bridge has done the work already.
+                            return;
+                        }
+                    }
+
+                    // Request the certificate from the Bridge
+                    string base64Cert = string.Empty;
+                    response = BridgeClient.MakeResourceGetRequest(CertificateAuthorityResourceName, null);
+
+                    string certificateAsBase64;
+                    if (response.TryGetValue(CertificateKeyResourceString, out certificateAsBase64))
+                    {
+                        certificateToInstall = new X509Certificate2(Convert.FromBase64String(certificateAsBase64));
+                    }
+                    else
+                    {
+                        StringBuilder sb = new StringBuilder();
+                        foreach (var pair in response)
+                        {
+                            sb.AppendFormat("{0}  {1} : {2}", Environment.NewLine, pair.Key, pair.Value);
```

Comment:
Nit: we should use the helper method to format a Dictionary<string, string> as a string for exceptions or tracing.